### PR TITLE
[Datasets] [Operator Fusion - E2E Mono-PR] [DO-NOT-MERGE] Add zero-copy operator fusion.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds support for zero-copy operator fusion, where we no longer materialize large blocks in between fused operator transformations. This is done by:
1. Changing the fundamental generated transform during planning from `Iterator[Block] -> Iterator[Block]` to `Iterator[DataT] -> Iterator[DataT]`, where `DataT = Union[Block DataBatch, Row]`.
2. Extracting all batching, formatting, and output buffering logic out into transform adapters.
3. Fusing all operators in a fusable linear chain at once, rather than fusing operators pairwise.
4. Creating the appropriate adapters between input blocks and the first operator, between fused operators, and between the last operator and the output blocks.

Most of this PR diff is adding full adapter coverage (both in functionality and in tests) for block-, batch-, and row-based transforms, where each of the transform data types need to be converted into the others without unnecessary copies/materializations. In addition, we have to handle the adapter between transform outputs and the eventual block outputs of the physical operator, which need to be buffered and split according to our dynamic block splitting policy.

## TODOs

- [x] Add e2e test coverage of the zero-copy semantics.
- [x] Add unit tests for adapters.
- [x] Add benchmark.
- [x] Add perf fixes.
- [x] Consolidate legacy planner path and new planner path (as much as possible).
- [x] Add copying and batch conversion metrics for operators, along with extended test coverage.
- [ ] Split into stacked PRs.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_execution_optimizer.py [hunk 1]
```diff
@@ -259,6 +263,115 @@ def test_read_map_batches_operator_fusion(ray_start_regular_shared, enable_optim
...
+
+    map_op2 = MapBatches(
+        map_op1,
+        check_batch,
+        batch_size=None,
+    )
+    logical_plan = LogicalPlan(map_op2)
+    physical_plan = planner.plan(logical_plan)
+    physical_plan = PhysicalOptimizer().optimize(physical_plan)
+    op = physical_plan.dag
```
Comment (ericl) [inline python/ray/data/tests/test_execution_optimizer.py:365]: Can we improve the testing / observability by emitting metrics on batch conversion operations and overheads? Then we can just assert the expected conversion in the metrics, e.g. `{"numpy_to_pandas_conversions": 5}` etc.

**Review by ericl — COMMENTED** [review]
Any microbenchmark results?

**Comment by clarkzinzow** [comment]
>Any microbenchmark results?

@c21 Do you have the benchmark/reproduction on-hand that highlighted this issue? Would love to validate this against that!

**Comment by c21** [comment]
@clarkzinzow - yeah we can start with map_batches benchmark in nightly test.

**Comment by clarkzinzow** [comment]
@c21 Just kicked off that nightly test, but I just realized that it won't use the new optimizer since it isn't enabled by default in master.

**Comment by c21** [comment]
@clarkzinzow - can you add `ray.data.context.DatasetContext.get_current().optimizer_enabled = True` inside `map_batches_benchmark.py` in this PR? So you can kick off the test?

File: python/ray/data/tests/test_execution_optimizer.py (hunk 1 above)
Comment (ericl) [reply python/ray/data/tests/test_execution_optimizer.py:365]: Should we do this?

File: python/ray/data/_internal/arrow_block.py [hunk 2]
```diff
@@ -224,7 +228,7 @@ def schema(self) -> "pyarrow.lib.Schema":
     def to_pandas(self) -> "pandas.DataFrame":
         from ray.air.util.data_batch_conversion import _cast_tensor_columns_to_ndarrays
 
-        df = self._table.to_pandas()
+        df = self._table.to_pandas(use_threads=False)
```
Comment (ericl) [inline python/ray/data/_internal/arrow_block.py:231]: Any particular reason for this / document the reason for this? Naively it seems like we'd want to use threads.

**Review by ericl — COMMENTED** [review]
Meta-comment: is it possible to extract a interface PR to review the high level strategy first, followed by others?

File: python/ray/data/_internal/arrow_block.py (hunk 2 above)
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_block.py:231]: @ericl As mentioned in standup and in the PR description, the plan is to decompose this PR into a stack, this PR is just the end-to-end mono-PR used for making sure that everything works and for benchmarking the end-state.

**Comment by clarkzinzow** [comment]
@ericl As mentioned in standup and in the PR description, the plan is to decompose this PR into a stack, this PR is just the end-to-end mono-PR used for making sure that everything works and for benchmarking the end-state.

File: python/ray/data/_internal/arrow_block.py (hunk 2 above)
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_block.py:231]: I didn't notice having any positive or negative effect on my benchmarking after disabling multithreaded Pandas conversion, so I thought I'd set this to `False` similar to how we disable multi-threaded Arrow I/O, and only enable it if it showed speedups in the benchmarking.

I think I'll probably set this back to the default for now when I split out the PRs, though.

**Pull request closed without merging**